pkg/orchestration/wiring/saml/plugin.go

```diff
@@ -0,0 +1,136 @@
+	ResourceType                  voyager.ResourceType = "SAML"
+	ResourcePrefix                                     = "SAML"
+)
+
+type autowiringOnlySpec struct {
+	Name        string `json:"name"`
+	Description string `json:"description"`
+	URL         string `json:"url"`
+}
+type finalSpec struct {
+	autowiringOnlySpec
+}
+type WiringPlugin struct {
+}
+
+func New() *WiringPlugin {
+	return &WiringPlugin{}
+}
+
+func (p *WiringPlugin) WireUp(resource *orch_v1.StateResource, context *wiringplugin.WiringContext) wiringplugin.WiringResult {
+	if resource.Type != ResourceType {
+		return &wiringplugin.WiringResultFailure{
+			Error: errors.Errorf("invalid resource type: %q", resource.Type),
+		}
+	}
+
+	serviceInstance, err := osb.ConstructServiceInstance(resource, clusterServiceClassExternalID, clusterServicePlanExternalID)
+	if err != nil {
+		return &wiringplugin.WiringResultFailure{
+			Error: err,
+		}
+	}
+
+	instanceParameters, external, retriable, err := instanceParameters(resource, context)
+	if err != nil {
+		return &wiringplugin.WiringResultFailure{
+			Error:            err,
+			IsExternalError:  external,
+			IsRetriableError: retriable,
+		}
+	}
+	serviceInstance.Spec.Parameters = &runtime.RawExtension{
+		Raw: instanceParameters,
+	}
+
+	instanceResourceName := wiringutil.ServiceInstanceResourceName(resource.Name)
+
+	smithResource := smith_v1.Resource{
+		Name:       instanceResourceName,
+		References: nil,
+		Spec: smith_v1.ResourceSpec{
+			Object: serviceInstance,
+		},
+	}
+
+	shapes, external, retriable, err := instanceShapes(&smithResource)
+	if err != nil {
+		return &wiringplugin.WiringResultFailure{
+			Error:            err,
+			IsExternalError:  external,
+			IsRetriableError: retriable,
+		}
+	}
+
+	return &wiringplugin.WiringResultSuccess{
+		Contract: wiringplugin.ResourceContract{
+			Shapes: shapes,
+		},
+		Resources: []smith_v1.Resource{smithResource},
+	}
+}
+
+func instanceShapes(smithResource *smith_v1.Resource) ([]wiringplugin.Shape, bool, bool, error) {
+	bindableEnvVarShape := knownshapes.NewBindableEnvironmentVariablesWithExcludeResourceName(smithResource.Name, ResourcePrefix, map[string]string{
+		"SAML_IDP_METADATA_URL": "data.saml_idp_metadata_url",
+		"IDP_METADATA_URL":      "data.idp_metadata_url",
+	}, true)
+	return []wiringplugin.Shape{
+		bindableEnvVarShape,
+	}, false, false, nil
+}
+
+func instanceParameters(resource *orch_v1.StateResource, context *wiringplugin.WiringContext) ([]byte, bool /* externalErr */, bool /* retriable */, error) {
+	if len(context.Dependencies) > 0 {
+		return nil, true, false, errors.New("saml resource should not have any dependencies")
+	}
+	if resource.Spec == nil {
+		return nil, true, false, errors.New("saml resource is missing user parameters")
+	}
+
+	var spec finalSpec
+
+	err := json.Unmarshal(resource.Spec.Raw, &spec)
+	if err != nil {
+		return nil, true, false, errors.WithStack(err)
```

Comment:
This is probably an internal error because json.Unmarshal shouldn't fail unless the resource spec is wrong, and we already check that via the CRD schema

Reply:
Yes, you are right.

Reply:
Oops. Sorry I didn't realize you commented on the same thing.

Reply:
You two commented the entirely opposite things 😛

Reply:
I think Yiyang is right, this should be an internal error.